### ISSUE TITLE
Bug: (fill in)

### ISSUE THREAD

**Issue by stephen-carter-at-sf** [eslint/eslint#19691]
### Environment

Node version: 23.6.1
npm version: 10.9.2
ESLint Package versions: 
```
    "@eslint/js": "^9.26.0",
    "eslint": "^9.26.0",
    "globals": "^16.0.0",
    "typescript-eslint": "^8.31.1"
```
Operating System: MacOs - Sequoia 15.4.1

### What parser are you using?

Default (Espree)

### What did you do?

isPathIgnored (which I believe just uses calculateConfigForFile) is not ignoring my files as I expect it to based on my simple configuration.

A few things I noticed:
- If I specify for a configuration to just use `files: ["**/*.js"]` for example then it does not ignore my .mjs files as well. 
  - Even if I also add in `ignores: ["**/*.mjs"]` it still doesn't ignore my .mjs files
- If I specify `files: ['**/*.ts"]` and `ignores: ['**/*.js']` then it doesn't always ignore my js files (depending on whether or not I provide cwd or not. Also it doesn't ignore my '.mjs' files again.

<details>
Here is a test to reproduce very easily:
```js
import {ESLint, Linter} from "eslint";
import {globalIgnores} from "eslint/config";
import js from "@eslint/js";
import ts from "typescript-eslint";

test1With('/some/path', '/some/path/dummy.js', false);       // [PASS]
test1With(undefined, '/some/path/dummy.js', false);          // [FAIL] - WHY? I'm still supplying an absolute path
test1With('/some/path/other', '/some/path/dummy.js', false); // [FAIL] - WHY? I'm still supplying an absolute path
test1With('/some/path', '/some/path/dummy.mjs', true);       // [FAIL] - WHY? I say that I only want *.js files ...
test1With(undefined, 'dummy.mjs', true);                     // [FAIL] - WHY? ... and I'm even ignoring *.mjs files
test1With('/some/path', '/some/path/dummy.ts', true);        // [PASS]
test1With(undefined, 'dummy.ts', true);                      // [PASS]

test2With('/some/path', '/some/path/dummy.js', true);        // [FAIL] - WHY? I say that I only want *.ts files ...
test2With(undefined, '/some/path/dummy.js', true);           // [PASS] - This seems inconsistent with test1
test2With('/some/path/other', '/some/path/dummy.js', true);  // [PASS]
test2With('/some/path', '/some/path/dummy.mjs', true);       // [FAIL] - WHY? I don't want .mjs files at all ...
test2With(undefined, 'dummy.mjs', true);                     // [FAIL] - WHY? ... ditto
test2With('/some/path', '/some/path/dummy.ts', false);       // [PASS]
test2With(undefined, 'dummy.ts', false);                     // [PASS]

async function test1With(cwdValue, file, expected) {
    const eslint = new ESLint({
        cwd: cwdValue,
        baseConfig: [{
            ... js.configs.recommended,
            files: ["**/*.js"],
            ignores: ["**/*.mjs"] // Even explicitely ignoring *.mjs files but it doesn't work
        }],
        overrideConfigFile: true
    });
    
    const ignored = await eslint.isPathIgnored(file);
    
    const header = `\nTest1: ${JSON.stringify({cwd: cwdValue, file: file})}`;
    printResult(`\nTest1: ${JSON.stringify({cwd: cwdValue, file: file})}`, ignored, expected);
}

async function test2With(cwdValue, file, expected) {
    const eslint = new ESLint({
        cwd: cwdValue,
        baseConfig: [
        {
            files: ["**/*.ts"],
            ignores: ["**/*.js"] // Even explicitely ignoring *.js files but it doesn't work
        },
        ...ts.configs.recommended],
        overrideConfigFile: true
    });
    
    const ignored = await eslint.isPathIgnored(file);
    
    printResult(`\nTest2: ${JSON.stringify({cwd: cwdValue, file: file})}`, ignored, expected);
}

function printResult(header, actual, expected) {
    if (actual != expected) {
        console.log(`${header}\n  [FAIL] isPathIgnored returned ${actual} but expected ${expected}`);
    } else {
        console.log(`${header}\n  [PASS] isPathIgnored returned ${actual} as expected`);
    }
}
```
</details>

### What did you expect to happen?

It seems like the behavior changed from ESLint 8 and the legacy configs to ESLint 9 and the flat configs. Basically, i would expect that if a user only supplies "files" then only the files that match that pattern shouldn't be ignored and all others should be ignored. Likewise, if a user only supplies "ignores" then all files are accepted except those that match the pattern given by ignores. Lastly, if both "files" and ignores are specified then it first only includes only the files that match the "files" pattern and then excludes from that set anything that matches the "ignores" pattern.

### What actually happened?

It seems like there are files that are not getting ignored when they should even when they are explicitly ignored or implicitly ignored.

See the details above.

### Link to Minimal Reproducible Example

https://github.com/stephen-carter-at-sf/eslint_bug_isPathIgnored

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by stephen-carter-at-sf** [comment]
If I put in a debug breakpoint while running test2 inside of the eslint/lib/config/config-loader.js at line 582 then I'm seeing that before it calls through to FlatConfigArray, the base config has exactly what I want. But after it calls FlatConfigArray it seems to add in additional config entires with .js, .mjs etc. Which seems wrong.

<img width="1562" alt="Image" src="https://github.com/user-attachments/assets/266a7e60-f015-4de8-ada1-630a5c633138" />

Not sure if this is the problem or not ... but found it interesting and unexpected.

**Comment by fasttime** [comment]
Thanks for the issue @stephen-carter-at-sf.

> It seems like the behavior changed from ESLint 8 and the legacy configs to ESLint 9 and the flat configs.

Definitely, the new config system is different from ESLint v8's in many ways.

> Basically, i would expect that if a user only supplies "files" then only the files that match that pattern shouldn't be ignored and all others should be ignored. Likewise, if a user only supplies "ignores" then all files are accepted except those that match the pattern given by ignores. Lastly, if both "files" and ignores are specified then it first only includes only the files that match the "files" pattern and then excludes from that set anything that matches the "ignores" pattern.
> 
> For example if I have
> 
> ```
>     const eslint = new ESLint({
>         cwd: '/some/path',
>         baseConfig: ts.configs.recommended.map(c => {
>             c.files = ["**/*.ts"]; // Forcefully putting "files" on the typescript config
>             return c;
>         }),
>         overrideConfigFile: true
>     });
> ```
> 
> then I would expect that `await eslint.isPathIgnored(file)` would return true for any .js file.

This is not how [`isPathIgnored`](https://eslint.org/docs/latest/integrate/nodejs-api#-eslintispathignoredfilepath) works. That method indicates if a file will be ignored by ESLint during linting. As per [the documentation](https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-and-ignores), files that match the patterns `**/*.js`, `**/*.cjs`, and `**/*.mjs` are linted by default. As you found out by yourself through debugging, ESLint prepends a default config to the user-provided config in order to enforce this behavior. It's also intentional that `ignores` patterns take priority over `files` patterns within the same configuration object, just as files outside the base path are always ignored. You are welcome to use the [discussions](https://github.com/eslint/eslint/discussions) if you need clarifications about the details of the API.
